File: loom_engine/src/main/scala/org/loom/scene/BrushConfig.scala
package org.loom.scene

/**
 * Configuration for brush-based rendering (BRUSHED mode).
 * Holds all parameters controlling how brush stamps are placed along polygon edges.
 */
class BrushConfig(
  var brushNames: Array[String],
  var drawMode: Int,
  var stampSpacing: Double,
  var spacingEasing: String,
  var followTangent: Boolean,
  var perpendicularJitterMin: Double,
  var perpendicularJitterMax: Double,
  var scaleMin: Double,
  var scaleMax: Double,
  var opacityMin: Double,
  var opacityMax: Double,
  var stampsPerFrame: Int,
  var agentCount: Int,
  var postCompletionMode: Int,
  var blurRadius: Int
) {

  /**
   * Scale all pixel-based values by the given factor (for quality scaling).
   * Called once during setup when quality > 1; BrushStampEngine then uses
   * the pre-scaled values directly without any further multiplication.
   */
  def scalePixelValues(factor: Double): Unit = {
    stampSpacing = stampSpacing * factor
    perpendicularJitterMin = perpendicularJitterMin * factor
    perpendicularJitterMax = perpendicularJitterMax * factor
    blurRadius = (blurRadius * factor).round.toInt
    stampsPerFrame = math.max(1, (stampsPerFrame * factor).round.toInt)
  }

  override def toString: String = {
    s"BrushConfig(brushes=${brushNames.mkString(",")}, mode=$drawMode, spacing=$stampSpacing)"
  }
}

object BrushConfig {
  // Draw modes
  val FULL_PATH: Int = 0
  val PROGRESSIVE: Int = 1

  // Post-completion modes (for progressive reveal)
  val HOLD: Int = 0
  val LOOP: Int = 1
  val PING_PONG: Int = 2

  def default(): BrushConfig = {
    BrushConfig(
      brushNames = Array("default.png"),
      drawMode = FULL_PATH,
      stampSpacing = 4.0,
      spacingEasing = "LINEAR",
      followTangent = true,
      perpendicularJitterMin = -2.0,
      perpendicularJitterMax = 2.0,
      scaleMin = 0.8,
      scaleMax = 1.2,
      opacityMin = 0.6,
      opacityMax = 1.0,
      stampsPerFrame = 10,
      agentCount = 1,
      postCompletionMode = HOLD,
      blurRadius = 0
    )
  }

  def apply(
    brushNames: Array[String],
    drawMode: Int,
    stampSpacing: Double,
    spacingEasing: String,
    followTangent: Boolean,
    perpendicularJitterMin: Double,
    perpendicularJitterMax: Double,
    scaleMin: Double,
    scaleMax: Double,
    opacityMin: Double,
    opacityMax: Double,
    stampsPerFrame: Int,
    agentCount: Int,
    postCompletionMode: Int,
    blurRadius: Int
  ): BrushConfig = {
    new BrushConfig(
      brushNames, drawMode, stampSpacing, spacingEasing, followTangent,
      perpendicularJitterMin, perpendicularJitterMax,
      scaleMin, scaleMax, opacityMin, opacityMax,
      stampsPerFrame, agentCount, postCompletionMode, blurRadius
    )
  }
}

File: loom_swift/Sources/LoomEngine/Rendering/BrushConfig.swift
import Foundation

// MARK: - MeanderConfig

/// Controls perpendicular displacement of brush paths for a hand-drawn look.
public struct MeanderConfig: Equatable, Codable, Sendable {
    public var enabled:                 Bool   = false
    public var amplitude:               Double = 8.0    // max displacement in canvas pixels
    public var frequency:               Double = 0.03   // noise control-points per pixel
    public var samples:                 Int    = 24     // sample points along each edge
    public var seed:                    Int    = 0      // 0 = auto; >0 = fixed seed
    public var animated:                Bool   = false
    public var animSpeed:               Double = 0.01
    public var scaleAlongPath:          Bool   = false
    public var scaleAlongPathFrequency: Double = 0.05
    public var scaleAlongPathRange:     Double = 0.4

    public init() {}
}

// MARK: - DrawMode / PostCompletionMode

public enum BrushDrawMode: String, Codable, Sendable {
    case fullPath    = "FULL_PATH"
    case progressive = "PROGRESSIVE"
}

public enum PostCompletionMode: String, Codable, Sendable {
    case hold     = "HOLD"
    case loop     = "LOOP"
    case pingPong = "PING_PONG"
}

// MARK: - BrushConfig

/// All parameters for `RendererMode.brushed` stamping.
///
/// Pixel-space values (`stampSpacing`, `perpendicularJitter*`, `meanderConfig.amplitude`)
/// are stored at the logical (quality=1) scale.  `scaled(by:)` returns a copy adjusted
/// for the project's `qualityMultiple`.
public struct BrushConfig: Equatable, Codable, Sendable {
    /// Filenames of brush PNG images relative to `<project>/brushes/`.
    public var brushNames:              [String]          = ["default.png"]
    public var drawMode:                BrushDrawMode     = .fullPath
    public var stampSpacing:            Double            = 4.0
    public var spacingEasing:           String            = "LINEAR"
    public var followTangent:           Bool              = true
    public var perpendicularJitterMin:  Double            = -2.0
    public var perpendicularJitterMax:  Double            = 2.0
    public var scaleMin:                Double            = 0.8
    public var scaleMax:                Double            = 1.2
    public var opacityMin:              Double            = 0.6
    public var opacityMax:              Double            = 1.0
    public var stampsPerFrame:          Int               = 10
    public var agentCount:              Int               = 1
    public var postCompletionMode:      PostCompletionMode = .hold
    public var blurRadius:              Int               = 0
    public var meander:                 MeanderConfig     = MeanderConfig()
    public var pressureSizeInfluence:   Double            = 0.0
    public var pressureAlphaInfluence:  Double            = 0.0

    public init() {}

    /// Return a copy with all pixel-space values multiplied by `factor`.
    ///
    /// `blurRadius` is scaled so the lookup key matches the pre-blurred image stored
    /// by `LoomEngine.preblurBrushImages` at `logicalRadius * qualityMultiple`.
    /// This mirrors Scala's `BrushConfig.scalePixelValues` exactly.
    public func scaled(by factor: Double) -> BrushConfig {
        guard factor != 1.0 else { return self }
        var c = self
        c.stampSpacing           *= factor
        c.perpendicularJitterMin *= factor
        c.perpendicularJitterMax *= factor
        c.blurRadius              = Int((Double(blurRadius) * factor).rounded())
        c.stampsPerFrame          = max(1, Int((Double(stampsPerFrame) * factor).rounded()))
        c.meander.amplitude      *= factor
        c.meander.frequency      /= factor   // keep visual density constant at higher res
        c.meander.scaleAlongPathFrequency /= factor
        c.meander.samples         = max(4, Int((Double(meander.samples) * factor).rounded()))
        return c
    }
}
